{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Load & Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/w1/mbcsphtx2xz1xqrls0v3d6r80000gn/T/ipykernel_69657/1976341293.py:2: DtypeWarning: Columns (8,10,12,16,17,18,19,20,21,22,23,24,26,31,32,37,38,45,46,47,48,49,50,51,52,53,60,61,66,67) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  iedb_raw = pd.read_csv('data/raw/iedb.csv')\n"
     ]
    }
   ],
   "source": [
    "vdjdb_raw = pd.read_csv('data/raw/vdjdb.tsv', sep='\\t')\n",
    "iedb_raw = pd.read_csv('data/raw/iedb.csv')\n",
    "tcr_train = pd.read_csv('data/net_tcr/train_ab_95_alphabeta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iedb_a_merge_curated = pd.merge(tcr_train['CDR3a'], iedb_a, how='inner', left_on='CDR3a', right_on='Chain 1 CDR3 Curated')\n",
    "iedb_a_merge_calculated = pd.merge(tcr_train['CDR3a'], iedb_a, how='inner', left_on='CDR3a', right_on='Chain 1 CDR3 Calculated')\n",
    "print('CDR3a IEDB curated:', iedb_a_merge_curated.shape)\n",
    "print('CDR3a IEDB calculated:', iedb_a_merge_calculated.shape)\n",
    "\n",
    "iedb_b_merge_curated = pd.merge(tcr_train['CDR3b'], iedb_b, how='inner', left_on='CDR3b', right_on='Chain 2 CDR3 Curated')\n",
    "iedb_b_merge_calculated = pd.merge(tcr_train['CDR3b'], iedb_b, how='inner', left_on='CDR3b', right_on='Chain 2 CDR3 Calculated')\n",
    "print('CDR3b IEDB curated:', iedb_b_merge_curated.shape)\n",
    "print('CDR3b IEDB calculated:', iedb_b_merge_calculated.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Net-TCR 2.0 paper:\n",
    "3859 CDRab chains 279 epitopes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VDJdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['complex.id', 'Gene', 'CDR3', 'V', 'J', 'Species', 'MHC A', 'MHC B',\n",
      "       'MHC class', 'Epitope', 'Epitope gene', 'Epitope species', 'Reference',\n",
      "       'Method', 'Meta', 'CDR3fix', 'Score'],\n",
      "      dtype='object')\n",
      "(58446, 17)\n"
     ]
    }
   ],
   "source": [
    "print(vdjdb_raw.columns)\n",
    "print(vdjdb_raw.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdj_cols = ['Gene', 'CDR3', 'V', 'J', 'Species', 'MHC A', 'MHC B',\n",
    "       'MHC class']\n",
    "vdjdb = vdjdb_raw[vdj_cols].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdjdb['CDR3'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# strip 1st and last AA from chain for lookup\n",
    "vdjdb['CDR3 strip'] = vdjdb['CDR3'].apply(lambda x: x[1:-1])\n",
    "\n",
    "# remove duplicates, keep last\n",
    "vdjdb = vdjdb.drop_duplicates(subset=['Gene', 'CDR3'], keep='last')\n",
    "vdjdb = vdjdb.astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>CDR3</th>\n",
       "      <th>V</th>\n",
       "      <th>J</th>\n",
       "      <th>Species</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>MHC B</th>\n",
       "      <th>MHC class</th>\n",
       "      <th>CDR3 strip</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAYRPPGTYKYIF</td>\n",
       "      <td>TRAV38-2/DV8*01</td>\n",
       "      <td>TRAJ40*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>AYRPPGTYKYI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSALASLNEQFF</td>\n",
       "      <td>TRBV14*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>ASSALASLNEQF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CIVRAPGRADMRF</td>\n",
       "      <td>TRAV26-1*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>IVRAPGRADMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>AVPSGAGSYQLT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAYTVLGNEKLTF</td>\n",
       "      <td>TRAV38-1*01</td>\n",
       "      <td>TRAJ48*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>AYTVLGNEKLT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58440</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASGSGGEHF</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>ASGSGGEH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58441</th>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVVGFSDGQKLLF</td>\n",
       "      <td>TRAV8-3*01</td>\n",
       "      <td>nan</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>AVVGFSDGQKLL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58442</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSELDGRL</td>\n",
       "      <td>TRBV2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>ASSELDGR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58444</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSPTDNF</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>ASSPTDN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58445</th>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSSYTQRGLDYTGELFF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>TRBJ2-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*24:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>ASSSYTQRGLDYTGELF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42789 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene                 CDR3                V           J      Species  \\\n",
       "2      TRA        CAYRPPGTYKYIF  TRAV38-2/DV8*01   TRAJ40*01  HomoSapiens   \n",
       "3      TRB       CASSALASLNEQFF        TRBV14*01  TRBJ2-1*01  HomoSapiens   \n",
       "8      TRA        CIVRAPGRADMRF      TRAV26-1*01   TRAJ43*01  HomoSapiens   \n",
       "9      TRA       CAVPSGAGSYQLTF        TRAV20*01   TRAJ28*01  HomoSapiens   \n",
       "10     TRA        CAYTVLGNEKLTF      TRAV38-1*01   TRAJ48*01  HomoSapiens   \n",
       "...    ...                  ...              ...         ...          ...   \n",
       "58440  TRB           CASGSGGEHF         TRBV9*01  TRBJ2-7*01  HomoSapiens   \n",
       "58441  TRA       CAVVGFSDGQKLLF       TRAV8-3*01         nan  HomoSapiens   \n",
       "58442  TRB           CASSELDGRL         TRBV2*01  TRBJ2-1*01  HomoSapiens   \n",
       "58444  TRB            CASSPTDNF         TRBV9*01  TRBJ2-3*01  HomoSapiens   \n",
       "58445  TRB  CASSSYTQRGLDYTGELFF        TRBV27*01  TRBJ2-2*01  HomoSapiens   \n",
       "\n",
       "             MHC A MHC B MHC class         CDR3 strip  \n",
       "2         HLA-B*08   B2M      MHCI        AYRPPGTYKYI  \n",
       "3         HLA-B*08   B2M      MHCI       ASSALASLNEQF  \n",
       "8         HLA-B*08   B2M      MHCI        IVRAPGRADMR  \n",
       "9         HLA-B*08   B2M      MHCI       AVPSGAGSYQLT  \n",
       "10        HLA-A*02   B2M      MHCI        AYTVLGNEKLT  \n",
       "...            ...   ...       ...                ...  \n",
       "58440  HLA-B*07:02   B2M      MHCI           ASGSGGEH  \n",
       "58441  HLA-B*07:02   B2M      MHCI       AVVGFSDGQKLL  \n",
       "58442  HLA-B*07:02   B2M      MHCI           ASSELDGR  \n",
       "58444  HLA-B*07:02   B2M      MHCI            ASSPTDN  \n",
       "58445  HLA-A*24:01   B2M      MHCI  ASSSYTQRGLDYTGELF  \n",
       "\n",
       "[42789 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdjdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VDJdb CDR3a: (20702, 9)\n",
      "VDJdb CDR3b: (22087, 9)\n"
     ]
    }
   ],
   "source": [
    "# split into alpha and beta tables\n",
    "vdjdb_a = vdjdb[vdjdb['Gene'] == 'TRA']\n",
    "vdjdb_b = vdjdb[vdjdb['Gene'] == 'TRB']\n",
    "\n",
    "print('VDJdb CDR3a:', vdjdb_a.shape)\n",
    "print('VDJdb CDR3b:', vdjdb_b.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDR3a unique sequences: 20702\n",
      "CDR3b unique sequences: 22086\n"
     ]
    }
   ],
   "source": [
    "## Convert VDJdb into CDR3 alpha and beta dictionaries\n",
    "\n",
    "vdjdb_a_idx = vdjdb_a.drop_duplicates(subset='CDR3 strip', keep='last')\n",
    "vdjdb_b_idx = vdjdb_b.drop_duplicates(subset='CDR3 strip', keep='last')\n",
    "\n",
    "vdjdb_a_idx = vdjdb_a_idx.set_index('CDR3 strip', drop=True)\n",
    "vdjdb_b_idx = vdjdb_b_idx.set_index('CDR3 strip', drop=True)\n",
    "\n",
    "vdjdb_a_dict = vdjdb_a_idx.to_dict('index')\n",
    "vdjdb_b_dict = vdjdb_b_idx.to_dict('index')\n",
    "\n",
    "print('CDR3a unique sequences:', len(vdjdb_a_dict.keys()))\n",
    "print('CDR3b unique sequences:', len(vdjdb_b_dict.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3a</th>\n",
       "      <th>CDR3b</th>\n",
       "      <th>peptide</th>\n",
       "      <th>partition</th>\n",
       "      <th>binder</th>\n",
       "      <th>cdra</th>\n",
       "      <th>va</th>\n",
       "      <th>ja</th>\n",
       "      <th>cdrb</th>\n",
       "      <th>vb</th>\n",
       "      <th>jb</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc</th>\n",
       "      <th>mhc_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADLQTNARLM</td>\n",
       "      <td>SGGQGMVDGYT</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARANNARLM</td>\n",
       "      <td>ASSLEYSPRPYEQY</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGGGSSNTGKLI</td>\n",
       "      <td>ASSLLLAGYNEQF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVRGTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AVRDTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16459</th>\n",
       "      <td>AVHRVGADKLI</td>\n",
       "      <td>ASSLGVWASGTDTQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16460</th>\n",
       "      <td>AASGPGANNLF</td>\n",
       "      <td>ASGRVGQETQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16461</th>\n",
       "      <td>AMKSTNAGKST</td>\n",
       "      <td>ASSVGQGVRSF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16462</th>\n",
       "      <td>AGLITQGGSEKLV</td>\n",
       "      <td>SVSRGPYEQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16463</th>\n",
       "      <td>AVTNNFGNEKLT</td>\n",
       "      <td>ASSYWPDYGYT</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16464 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3a            CDR3b    peptide  partition  binder  cdra  va  \\\n",
       "0         ADLQTNARLM      SGGQGMVDGYT  NLVPMVATV          5       1   NaN NaN   \n",
       "1         AARANNARLM   ASSLEYSPRPYEQY  NLVPMVATV          5       1   NaN NaN   \n",
       "2       AGGGSSNTGKLI    ASSLLLAGYNEQF  NLVPMVATV          5       1   NaN NaN   \n",
       "3         AVRGTNARLM        ASSVVTEAF  NLVPMVATV          5       1   NaN NaN   \n",
       "4         AVRDTNARLM        ASSVVTEAF  NLVPMVATV          5       1   NaN NaN   \n",
       "...              ...              ...        ...        ...     ...   ...  ..   \n",
       "16459    AVHRVGADKLI  ASSLGVWASGTDTQY  GILGFVFTL          5       0   NaN NaN   \n",
       "16460    AASGPGANNLF      ASGRVGQETQY  GILGFVFTL          5       0   NaN NaN   \n",
       "16461    AMKSTNAGKST      ASSVGQGVRSF  GILGFVFTL          5       0   NaN NaN   \n",
       "16462  AGLITQGGSEKLV       SVSRGPYEQY  GILGFVFTL          5       0   NaN NaN   \n",
       "16463   AVTNNFGNEKLT      ASSYWPDYGYT  GILGFVFTL          5       0   NaN NaN   \n",
       "\n",
       "       ja  cdrb  vb  jb  species  mhc  mhc_class  \n",
       "0     NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "1     NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "2     NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "3     NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "4     NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "...    ..   ...  ..  ..      ...  ...        ...  \n",
       "16459 NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "16460 NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "16461 NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "16462 NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "16463 NaN   NaN NaN NaN      NaN  NaN        NaN  \n",
       "\n",
       "[16464 rows x 14 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcr_add_cols = ['CDR3a', 'CDR3b', 'peptide', 'partition', 'binder']\n",
    "vdj_add_cols = ['cdra', 'va', 'ja', 'cdrb', 'vb', 'jb', 'species', 'mhc', 'mhc_class']\n",
    "tcr_train_geom = tcr_train.reindex(columns=tcr_add_cols+vdj_add_cols)\n",
    "tcr_train_geom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cdr3a absent: 12518\n",
      "cdr3b absent: 12518\n"
     ]
    }
   ],
   "source": [
    "## Search for CDR3 chains in TCR training data\n",
    "\n",
    "vdjdb_absent = {'cdr3a': list(), 'cdr3b': list()}\n",
    "for idx in tcr_train_geom.index:\n",
    "    try:\n",
    "        vdj_data_a = vdjdb_a_dict[tcr_train_geom.iloc[idx]['CDR3a']]\n",
    "        vdj_data_b = vdjdb_b_dict[tcr_train_geom.iloc[idx]['CDR3b']]\n",
    "        assert vdj_data_a['Species'] == vdj_data_b['Species'] # check that species match\n",
    "        assert vdj_data_a['MHC A'] == vdj_data_b['MHC A'] # check that HLA structures match\n",
    "        assert vdj_data_a['MHC class'] == vdj_data_b['MHC class'] # check that MHC classes match\n",
    "        tcr_train_geom.loc[idx, 'cdra'] = vdj_data_a['CDR3']\n",
    "        tcr_train_geom.loc[idx, 'va'] = vdj_data_a['V']\n",
    "        tcr_train_geom.loc[idx, 'ja'] = vdj_data_a['J']\n",
    "        tcr_train_geom.loc[idx, 'cdrb'] = vdj_data_b['CDR3']\n",
    "        tcr_train_geom.loc[idx, 'vb'] = vdj_data_b['V']\n",
    "        tcr_train_geom.loc[idx, 'jb'] = vdj_data_b['J']\n",
    "        tcr_train_geom.loc[idx, 'species'] = vdj_data_a['Species']\n",
    "        tcr_train_geom.loc[idx, 'mhc'] = vdj_data_a['MHC A']\n",
    "        tcr_train_geom.loc[idx, 'mhc_class'] = vdj_data_a['MHC class']\n",
    "    except:\n",
    "        vdjdb_absent['cdr3a'].append(tcr_train_geom.iloc[idx]['CDR3a'])\n",
    "        vdjdb_absent['cdr3b'].append(tcr_train_geom.iloc[idx]['CDR3b'])\n",
    "        \n",
    "print('cdr3a absent:', len(vdjdb_absent['cdr3a']))\n",
    "print('cdr3b absent:', len(vdjdb_absent['cdr3b']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3a</th>\n",
       "      <th>CDR3b</th>\n",
       "      <th>peptide</th>\n",
       "      <th>partition</th>\n",
       "      <th>binder</th>\n",
       "      <th>cdra</th>\n",
       "      <th>va</th>\n",
       "      <th>ja</th>\n",
       "      <th>cdrb</th>\n",
       "      <th>vb</th>\n",
       "      <th>jb</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc</th>\n",
       "      <th>mhc_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADLQTNARLM</td>\n",
       "      <td>SGGQGMVDGYT</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CADLQTNARLMF</td>\n",
       "      <td>TRAV5*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CSGGQGMVDGYTF</td>\n",
       "      <td>TRBV29-1*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARANNARLM</td>\n",
       "      <td>ASSLEYSPRPYEQY</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAARANNARLMF</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSLEYSPRPYEQYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGGGSSNTGKLI</td>\n",
       "      <td>ASSLLLAGYNEQF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGGGSSNTGKLIF</td>\n",
       "      <td>TRAV8-3*01</td>\n",
       "      <td>TRAJ37*01</td>\n",
       "      <td>CASSLLLAGYNEQFF</td>\n",
       "      <td>TRBV11-2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVRGTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAVRGTNARLMF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSVVTEAFF</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AVRDTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAVRDTNARLMF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSVVTEAFF</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16459</th>\n",
       "      <td>AVHRVGADKLI</td>\n",
       "      <td>ASSLGVWASGTDTQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16460</th>\n",
       "      <td>AASGPGANNLF</td>\n",
       "      <td>ASGRVGQETQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16461</th>\n",
       "      <td>AMKSTNAGKST</td>\n",
       "      <td>ASSVGQGVRSF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16462</th>\n",
       "      <td>AGLITQGGSEKLV</td>\n",
       "      <td>SVSRGPYEQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16463</th>\n",
       "      <td>AVTNNFGNEKLT</td>\n",
       "      <td>ASSYWPDYGYT</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>CAVTNNFGNEKLTF</td>\n",
       "      <td>TRAV36/DV7*01</td>\n",
       "      <td>TRAJ48*01</td>\n",
       "      <td>CASSYWPDYGYTF</td>\n",
       "      <td>TRBV6-3*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16464 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3a            CDR3b    peptide  partition  binder  \\\n",
       "0         ADLQTNARLM      SGGQGMVDGYT  NLVPMVATV          5       1   \n",
       "1         AARANNARLM   ASSLEYSPRPYEQY  NLVPMVATV          5       1   \n",
       "2       AGGGSSNTGKLI    ASSLLLAGYNEQF  NLVPMVATV          5       1   \n",
       "3         AVRGTNARLM        ASSVVTEAF  NLVPMVATV          5       1   \n",
       "4         AVRDTNARLM        ASSVVTEAF  NLVPMVATV          5       1   \n",
       "...              ...              ...        ...        ...     ...   \n",
       "16459    AVHRVGADKLI  ASSLGVWASGTDTQY  GILGFVFTL          5       0   \n",
       "16460    AASGPGANNLF      ASGRVGQETQY  GILGFVFTL          5       0   \n",
       "16461    AMKSTNAGKST      ASSVGQGVRSF  GILGFVFTL          5       0   \n",
       "16462  AGLITQGGSEKLV       SVSRGPYEQY  GILGFVFTL          5       0   \n",
       "16463   AVTNNFGNEKLT      ASSYWPDYGYT  GILGFVFTL          5       0   \n",
       "\n",
       "                 cdra             va         ja              cdrb  \\\n",
       "0        CADLQTNARLMF       TRAV5*01  TRAJ31*01     CSGGQGMVDGYTF   \n",
       "1        CAARANNARLMF  TRAV29/DV5*01  TRAJ31*01  CASSLEYSPRPYEQYF   \n",
       "2      CAGGGSSNTGKLIF     TRAV8-3*01  TRAJ37*01   CASSLLLAGYNEQFF   \n",
       "3        CAVRGTNARLMF       TRAV3*01  TRAJ31*01       CASSVVTEAFF   \n",
       "4        CAVRDTNARLMF       TRAV3*01  TRAJ31*01       CASSVVTEAFF   \n",
       "...               ...            ...        ...               ...   \n",
       "16459             NaN            NaN        NaN               NaN   \n",
       "16460             NaN            NaN        NaN               NaN   \n",
       "16461             NaN            NaN        NaN               NaN   \n",
       "16462             NaN            NaN        NaN               NaN   \n",
       "16463  CAVTNNFGNEKLTF  TRAV36/DV7*01  TRAJ48*01     CASSYWPDYGYTF   \n",
       "\n",
       "                vb          jb      species          mhc mhc_class  \n",
       "0      TRBV29-1*01  TRBJ1-2*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "1       TRBV7-9*01  TRBJ2-7*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "2      TRBV11-2*01  TRBJ2-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "3      TRBV12-3*01  TRBJ1-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "4      TRBV12-3*01  TRBJ1-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "...            ...         ...          ...          ...       ...  \n",
       "16459          NaN         NaN          NaN          NaN       NaN  \n",
       "16460          NaN         NaN          NaN          NaN       NaN  \n",
       "16461          NaN         NaN          NaN          NaN       NaN  \n",
       "16462          NaN         NaN          NaN          NaN       NaN  \n",
       "16463   TRBV6-3*01  TRBJ1-2*01  HomoSapiens  HLA-A*03:01      MHCI  \n",
       "\n",
       "[16464 rows x 14 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcr_train_geom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IEDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Group Receptor ID', 'Receptor ID', 'Reference IRI', 'Epitope IRI',\n",
      "       'Description', 'Antigen', 'Organism', 'Response Type', 'Assay IDs',\n",
      "       'MHC Allele Names', 'Reference Name', 'STABLE ID', 'Synonyms',\n",
      "       'Receptor Type', 'Chain 1 Type', 'Chain 1 Species',\n",
      "       'Chain 1 Nucleotide', 'Curated Chain 1 V Gene',\n",
      "       'Calculated Chain 1 V Gene', 'Curated Chain 1 D Gene',\n",
      "       'Calculated Chain 1 D Gene', 'Curated Chain 1 J Gene',\n",
      "       'Calculated Chain 1 J Gene', 'Chain 1 Full Sequence',\n",
      "       'Chain 1 Accession', 'Chain 1 CDR3 Curated', 'Chain 1 CDR3 Calculated',\n",
      "       'Chain 1 CDR3 Start Curated', 'Chain 1 CDR3 End Curated',\n",
      "       'Chain 1 CDR3 Start Calculated', 'Chain 1 CDR3 End Calculated',\n",
      "       'Chain 1 CDR1 Curated', 'Chain 1 CDR1 Calculated',\n",
      "       'Chain 1 CDR1 Start Curated', 'Chain 1 CDR1 End Curated',\n",
      "       'Chain 1 CDR1 Start Calculated', 'Chain 1 CDR1 End Calculated',\n",
      "       'Chain 1 CDR2 Curated', 'Chain 1 CDR2 Calculated',\n",
      "       'Chain 1 CDR2 Start Curated', 'Chain 1 CDR2 End Curated',\n",
      "       'Chain 1 CDR2 Start Calculated', 'Chain 1 CDR2 End Calculated',\n",
      "       'Chain 2 Type', 'Chain 2 Species', 'Chain 2 Nucleotide',\n",
      "       'Curated Chain 2 V Gene', 'Calculated Chain 2 V Gene',\n",
      "       'Curated Chain 2 D Gene', 'Calculated Chain 2 D Gene',\n",
      "       'Curated Chain 2 J Gene', 'Calculated Chain 2 J Gene',\n",
      "       'Chain 2 Full Sequence', 'Chain 2 Accession', 'Chain 2 CDR3 Curated',\n",
      "       'Chain 2 CDR3 Calculated', 'Chain 2 CDR3 Start Curated',\n",
      "       'Chain 2 CDR3 End Curated', 'Chain 2 CDR3 Start Calculated',\n",
      "       'Chain 2 CDR3 End Calculated', 'Chain 2 CDR1 Curated',\n",
      "       'Chain 2 CDR1 Calculated', 'Chain 2 CDR1 Start Curated',\n",
      "       'Chain 2 CDR1 End Curated', 'Chain 2 CDR1 Start Calculated',\n",
      "       'Chain 2 CDR1 End Calculated', 'Chain 2 CDR2 Curated',\n",
      "       'Chain 2 CDR2 Calculated', 'Chain 2 CDR2 Start Curated',\n",
      "       'Chain 2 CDR2 End Curated', 'Chain 2 CDR2 Start Calculated',\n",
      "       'Chain 2 CDR2 End Calculated'],\n",
      "      dtype='object')\n",
      "(23040, 72)\n"
     ]
    }
   ],
   "source": [
    "print(iedb_raw.columns)\n",
    "print(iedb_raw.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22771, 72)\n"
     ]
    }
   ],
   "source": [
    "subset = ['Chain 1 CDR3 Curated', 'Chain 2 CDR3 Curated']\n",
    "iedb = iedb_raw.dropna(subset=subset)\n",
    "iedb = iedb.astype('str')\n",
    "print(iedb.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22771, 43)\n",
      "(22771, 43)\n"
     ]
    }
   ],
   "source": [
    "iedb_base_cols = ['Group Receptor ID', 'Receptor ID', 'Reference IRI', 'Epitope IRI',\n",
    "       'Description', 'Antigen', 'Organism', 'Response Type', 'Assay IDs',\n",
    "       'MHC Allele Names', 'Reference Name', 'STABLE ID', 'Synonyms',\n",
    "       'Receptor Type']\n",
    "\n",
    "iedb_a_cols = [col for col in iedb.columns if \"Chain 1\" in col]\n",
    "iedb_b_cols = [col for col in iedb.columns if \"Chain 2\" in col]\n",
    "\n",
    "iedb_a = iedb[iedb_base_cols + iedb_a_cols]\n",
    "iedb_b = iedb[iedb_base_cols + iedb_b_cols]\n",
    "\n",
    "print(iedb_a.shape)\n",
    "print(iedb_b.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDR3a unique sequences: 17054\n",
      "CDR3b unique sequences: 18051\n"
     ]
    }
   ],
   "source": [
    "iedb_a_idx = iedb_a.drop_duplicates(subset='Chain 1 CDR3 Curated', keep='last')\n",
    "iedb_b_idx = iedb_b.drop_duplicates(subset='Chain 2 CDR3 Curated', keep='last')\n",
    "\n",
    "iedb_a_idx = iedb_a_idx.set_index('Chain 1 CDR3 Curated', drop=True)\n",
    "iedb_b_idx = iedb_b_idx.set_index('Chain 2 CDR3 Curated', drop=True)\n",
    "\n",
    "iedb_a_dict = iedb_a_idx.to_dict('index')\n",
    "iedb_b_dict = iedb_b_idx.to_dict('index')\n",
    "\n",
    "print('CDR3a unique sequences:', len(iedb_a_dict.keys()))\n",
    "print('CDR3b unique sequences:', len(iedb_b_dict.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cdr3a absent: 16464\n",
      "cdr3b absent: 16464\n"
     ]
    }
   ],
   "source": [
    "iedb_absent = {'cdr3a': list(), 'cdr3b': list()}\n",
    "for idx in tcr_train_geom.index:\n",
    "    try:\n",
    "        iedb_data_a = iedb_a_dict[tcr_train_geom.iloc[idx]['CDR3a']]\n",
    "        iedb_data_b = iedb_b_dict[tcr_train_geom.iloc[idx]['CDR3b']]\n",
    "        assert iedb_data_a['Organism'] == iedb_data_b['Organism'] # check that species match\n",
    "        assert iedb_data_a['MHC Allele Names'] == iedb_data_b['MHC Allele Names'] # check that HLA structures match\n",
    "        # assert iedb_data_a['MHC class'] == iedb_data_b['MHC class'] # check that MHC classes match\n",
    "        tcr_train_geom.loc[idx, 'cdra'] = iedb_data_a['Chain 1 CDR3 Curated']\n",
    "        tcr_train_geom.loc[idx, 'va'] = iedb_data_a['Curated Chain 1 V Gene']\n",
    "        tcr_train_geom.loc[idx, 'ja'] = iedb_data_a['Curated Chain 1 J Gene']\n",
    "        tcr_train_geom.loc[idx, 'cdrb'] = iedb_data_b['Chain 2 CDR3 Curated']\n",
    "        tcr_train_geom.loc[idx, 'vb'] = iedb_data_b['Curated Chain 2 V Gene']\n",
    "        tcr_train_geom.loc[idx, 'jb'] = iedb_data_b['Curated Chain 2 J Gene']\n",
    "        tcr_train_geom.loc[idx, 'species'] = iedb_data_a['Organism']\n",
    "        tcr_train_geom.loc[idx, 'mhca'] = iedb_data_a['MHC Allele Names']\n",
    "        # tcr_train_geom.loc[idx, 'mhcb'] = iedb_data_a['MHC B']\n",
    "        # tcr_train_geom.loc[idx, 'mhc_class'] = iedb_data_a['MHC class']\n",
    "    except:\n",
    "        iedb_absent['cdr3a'].append(tcr_train_geom.iloc[idx]['CDR3a'])\n",
    "        iedb_absent['cdr3b'].append(tcr_train_geom.iloc[idx]['CDR3b'])\n",
    "        \n",
    "print('cdr3a absent:', len(iedb_absent['cdr3a']))\n",
    "print('cdr3b absent:', len(iedb_absent['cdr3b']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3a</th>\n",
       "      <th>CDR3b</th>\n",
       "      <th>peptide</th>\n",
       "      <th>partition</th>\n",
       "      <th>binder</th>\n",
       "      <th>cdra</th>\n",
       "      <th>va</th>\n",
       "      <th>ja</th>\n",
       "      <th>cdrb</th>\n",
       "      <th>vb</th>\n",
       "      <th>jb</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc</th>\n",
       "      <th>mhc_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADLQTNARLM</td>\n",
       "      <td>SGGQGMVDGYT</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CADLQTNARLMF</td>\n",
       "      <td>TRAV5*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CSGGQGMVDGYTF</td>\n",
       "      <td>TRBV29-1*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARANNARLM</td>\n",
       "      <td>ASSLEYSPRPYEQY</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAARANNARLMF</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSLEYSPRPYEQYF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGGGSSNTGKLI</td>\n",
       "      <td>ASSLLLAGYNEQF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAGGGSSNTGKLIF</td>\n",
       "      <td>TRAV8-3*01</td>\n",
       "      <td>TRAJ37*01</td>\n",
       "      <td>CASSLLLAGYNEQFF</td>\n",
       "      <td>TRBV11-2*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVRGTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAVRGTNARLMF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSVVTEAFF</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AVRDTNARLM</td>\n",
       "      <td>ASSVVTEAF</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>CAVRDTNARLMF</td>\n",
       "      <td>TRAV3*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>CASSVVTEAFF</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16459</th>\n",
       "      <td>AVHRVGADKLI</td>\n",
       "      <td>ASSLGVWASGTDTQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16460</th>\n",
       "      <td>AASGPGANNLF</td>\n",
       "      <td>ASGRVGQETQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16461</th>\n",
       "      <td>AMKSTNAGKST</td>\n",
       "      <td>ASSVGQGVRSF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16462</th>\n",
       "      <td>AGLITQGGSEKLV</td>\n",
       "      <td>SVSRGPYEQY</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16463</th>\n",
       "      <td>AVTNNFGNEKLT</td>\n",
       "      <td>ASSYWPDYGYT</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>CAVTNNFGNEKLTF</td>\n",
       "      <td>TRAV36/DV7*01</td>\n",
       "      <td>TRAJ48*01</td>\n",
       "      <td>CASSYWPDYGYTF</td>\n",
       "      <td>TRBV6-3*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>MHCI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16464 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               CDR3a            CDR3b    peptide  partition  binder  \\\n",
       "0         ADLQTNARLM      SGGQGMVDGYT  NLVPMVATV          5       1   \n",
       "1         AARANNARLM   ASSLEYSPRPYEQY  NLVPMVATV          5       1   \n",
       "2       AGGGSSNTGKLI    ASSLLLAGYNEQF  NLVPMVATV          5       1   \n",
       "3         AVRGTNARLM        ASSVVTEAF  NLVPMVATV          5       1   \n",
       "4         AVRDTNARLM        ASSVVTEAF  NLVPMVATV          5       1   \n",
       "...              ...              ...        ...        ...     ...   \n",
       "16459    AVHRVGADKLI  ASSLGVWASGTDTQY  GILGFVFTL          5       0   \n",
       "16460    AASGPGANNLF      ASGRVGQETQY  GILGFVFTL          5       0   \n",
       "16461    AMKSTNAGKST      ASSVGQGVRSF  GILGFVFTL          5       0   \n",
       "16462  AGLITQGGSEKLV       SVSRGPYEQY  GILGFVFTL          5       0   \n",
       "16463   AVTNNFGNEKLT      ASSYWPDYGYT  GILGFVFTL          5       0   \n",
       "\n",
       "                 cdra             va         ja              cdrb  \\\n",
       "0        CADLQTNARLMF       TRAV5*01  TRAJ31*01     CSGGQGMVDGYTF   \n",
       "1        CAARANNARLMF  TRAV29/DV5*01  TRAJ31*01  CASSLEYSPRPYEQYF   \n",
       "2      CAGGGSSNTGKLIF     TRAV8-3*01  TRAJ37*01   CASSLLLAGYNEQFF   \n",
       "3        CAVRGTNARLMF       TRAV3*01  TRAJ31*01       CASSVVTEAFF   \n",
       "4        CAVRDTNARLMF       TRAV3*01  TRAJ31*01       CASSVVTEAFF   \n",
       "...               ...            ...        ...               ...   \n",
       "16459             NaN            NaN        NaN               NaN   \n",
       "16460             NaN            NaN        NaN               NaN   \n",
       "16461             NaN            NaN        NaN               NaN   \n",
       "16462             NaN            NaN        NaN               NaN   \n",
       "16463  CAVTNNFGNEKLTF  TRAV36/DV7*01  TRAJ48*01     CASSYWPDYGYTF   \n",
       "\n",
       "                vb          jb      species          mhc mhc_class  \n",
       "0      TRBV29-1*01  TRBJ1-2*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "1       TRBV7-9*01  TRBJ2-7*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "2      TRBV11-2*01  TRBJ2-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "3      TRBV12-3*01  TRBJ1-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "4      TRBV12-3*01  TRBJ1-1*01  HomoSapiens  HLA-A*02:01      MHCI  \n",
       "...            ...         ...          ...          ...       ...  \n",
       "16459          NaN         NaN          NaN          NaN       NaN  \n",
       "16460          NaN         NaN          NaN          NaN       NaN  \n",
       "16461          NaN         NaN          NaN          NaN       NaN  \n",
       "16462          NaN         NaN          NaN          NaN       NaN  \n",
       "16463   TRBV6-3*01  TRBJ1-2*01  HomoSapiens  HLA-A*03:01      MHCI  \n",
       "\n",
       "[16464 rows x 14 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcr_train_geom"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge VDJ to Net TCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdj_a_merge = pd.merge(tcr_train, vdjdb_a, how='left', left_on='CDR3a', right_on='CDR3')\n",
    "vdj_a_merge_strip = pd.merge(tcr_train, vdjdb_a, how='left', left_on='CDR3a', right_on='CDR3 strip')\n",
    "print('CDR3a VDJdb:', vdj_a_merge.shape)\n",
    "print('CDR3a VDJdb stripped:', vdj_a_merge_strip.shape)\n",
    "\n",
    "vdj_b_merge = pd.merge(tcr_train, vdjdb_b, how='left', left_on='CDR3b', right_on='CDR3')\n",
    "vdj_b_merge_strip = pd.merge(tcr_train, vdjdb_b, how='left', left_on='CDR3b', right_on='CDR3 strip')\n",
    "print('CDR3b VDJdb:', vdj_b_merge.shape)\n",
    "print('CDR3b VDJdb stripped:', vdj_b_merge_strip.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdj_merge = pd.merge(vdj_a_merge_strip, vdj_b_merge_strip, how='inner', on=['CDR3a', 'CDR3b'])\n",
    "vdj_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdj_merge = vdj_merge.drop_duplicates(subset=['CDR3_x', 'CDR3_y'], keep='last')\n",
    "assert (vdj_merge['partition_x']==vdj_merge['partition_y']).all()\n",
    "assert (vdj_merge['binder_x']==vdj_merge['binder_y']).all()\n",
    "assert (vdj_merge['peptide_x']==vdj_merge['peptide_y']).all()\n",
    "vdj_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('geo_lightning')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "73e2ca23702b1a0a44eb382a701e0753c1892c2a9c18bb9bb5c5b2fcf6d815fb"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
